src/main/kotlin/org/rust/lang/core/psi/ext/RsQualifiedNamedElement.kt

```diff
@@ -356,29 +356,37 @@ data class RsQualifiedName private constructor(
         }
 
         private fun RsQualifiedNamedElement.toParentItem(): Item? {
-            val name = ((this as? RsMod)?.modName ?: name) ?: return null
+            val name = when {
+                this is RsFunction && isProcMacroDef -> procMacroName
+                this is RsMod && isCrateRoot -> containingCrate?.normName
+                this is RsMod -> modName
+                else -> name
+            } ?: return null
```

Comment:
I would suggest move `name` declaration after `itemType` so that we don't calculate `isProcMacroDef` and `isCrateRoot` twice

```suggestion
            val name = when (itemType) {
                DERIVE -> (this as? RsFunction)?.procMacroName
                CRATE -> containingCrate?.normName
                MOD -> (this as? RsMod)?.modName
                else -> name
            } ?: return null
```